### PR TITLE
Implement Method to Get All Test Sessions

### PR DESCRIPTION
**Pull request by mahadzaryab1**
## Notion ticket link

[Get all test sessions](https://www.notion.so/uwblueprintexecs/Get-all-test-sessions-b77ba97e037341ac862deff161b65a53)

## Implementation description
* Added a new function `getAllTestSessions` to fetch all the test sessions from the database
* Fixed some bugs/syntax errors

## Steps to test
1.

## What should reviewers focus on?
* 

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/models/testSession.model.ts [hunk 1]
```diff
@@ -13,12 +13,12 @@ export interface Result {
    * a list corresponding to the question list with each field indicating
    * the student's answer
    */
-  answers: [number];
+  answers: number[];
```
Comment (mahadzaryab1) [inline backend/models/testSession.model.ts:16]: this and the following changes fix a syntax error that I made while writing these models

File: backend/services/implementations/__tests__/testSessionService.test.ts [hunk 2]
```diff
@@ -25,10 +29,14 @@ describe("mongo testSessionService", (): void => {
   it("createTestSession", async () => {
     const res = await testSessionService.createTestSession(mockTestSession);
 
-    expect(res.id).not.toBeNull();
-    expect(res).toMatchObject({
-      id: res.id,
-      ...mockTestSession,
-    });
+    // TODO: uncomment when results are added to test session response object
+    // assertResponseMatchesExpected(mockTestSession, res);
```
Comment (mahadzaryab1) [inline backend/services/implementations/__tests__/testSessionService.test.ts]: There's missing logic in `createTestSession` that doesn't map the Results when returning the response object. Should create a separate ticket for it.

File: backend/testUtils/testSession.ts [hunk 3]
```diff
@@ -1,10 +1,49 @@
...
 
-export default mockTestSession;
+export const assertResponseMatchesExpected = (
+  expected: TestSessionRequestDTO,
+  result: TestSessionResponseDTO,
+): void => {
+  expect(result.id).not.toBeNull();
+  expect(result.test.toString()).toEqual(expected.test);
+  expect(result.teacher.toString()).toEqual(expected.teacher);
+  expect(result.school.toString()).toEqual(expected.school);
```
Comment (mahadzaryab1) [inline backend/testUtils/testSession.ts:31]: did we decide on whether we want these to return the actual objects rather than just the IDs? if so, we can create a new ticket for that and do it in a separate PR

File: backend/services/implementations/__tests__/testSessionService.test.ts (hunk 2 above)
Comment (carissa-tang) [reply backend/services/implementations/__tests__/testSessionService.test.ts]: i omitted the `Results` field because it will be empty when a test session is first created, and was thinking we could add that attribute when a test session is updated. or do you think it would be better to have an empty `Results` field? I was having some trouble figuring out the typing for that

File: backend/testUtils/testSession.ts (hunk 3 above)
Comment (carissa-tang) [reply backend/testUtils/testSession.ts:31]: i believe we are returning the actual objects - i was thinking we could use `getTestById`, `getTeacherById` and `getSchoolById` once they are implemented

**Review by carissa-tang — APPROVED** [review]
lgtm! will need to create new tickets to update the mapping of ids to objects and add `result` to response of `createTestSession()`

File: backend/services/implementations/__tests__/testSessionService.test.ts (hunk 2 above)
Comment (mahadzaryab1) [reply backend/services/implementations/__tests__/testSessionService.test.ts]: ah i see - that's fine! i changed the test up to reflect this

**Review by carissa-tang — APPROVED** [review]

File: backend/models/testSession.model.ts (hunk 1 above)
Comment (joyce-shi) [reply backend/models/testSession.model.ts:16]: nice

File: backend/testUtils/testSession.ts [hunk 4]
```diff
@@ -1,10 +1,54 @@
...
+  ResultRequestDTO,
+  TestSessionRequestDTO,
+  TestSessionResponseDTO,
+} from "../services/interfaces/testSessionService";
+
+export const testResult: ResultRequestDTO = {
+  student: "some-student-name",
+  score: 25,
+  answers: [10, 11],
+  breakdown: [false, true, true],
```
Comment (joyce-shi) [inline backend/testUtils/testSession.ts]: wait - how can breakdown.length be > than answer.length lol

File: backend/services/implementations/testSessionService.ts [hunk 5]
```diff
@@ -38,6 +38,53 @@ class TestSessionService implements ITestSessionService {
...
+      Logger.error(
+        `Failed to get test sessions. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+
+    return testSessionDtos;
+  }
+
+  private async mapTestSessionsToTestSessionDTOs(
```
Comment (joyce-shi) [inline backend/services/implementations/testSessionService.ts:60]: not sure if you know the answer to this - but how does mongo handle large requests? i.e say we have thousands of test sessions, if we make a request for "all test sessions", how does it behave?

File: backend/services/interfaces/testSessionService.ts [hunk 6]
```diff
@@ -43,13 +43,49 @@ export interface TestSessionResponseDTO {
...
    * */
-  results?: [Result];
+  results?: ResultResponseDTO[];
   /** the code that students can use to access the test when it is live */
   accessCode: string;
   /** the time when the test session is started by teacher */
   startTime: Date;
 }
 
+export interface ResultRequestDTO {
```
Comment (joyce-shi) [inline backend/services/interfaces/testSessionService.ts:53]: why do we add `ResultRequestDTO` and `ResultResponseDTO`? is it for the sake of modularity (since we are using it as a DTO) or is it strictly necessary for another reason?

File: backend/testUtils/testSession.ts (hunk 3 above)
Comment (joyce-shi) [reply backend/testUtils/testSession.ts:31]: @carissa-tang sorry if I'm misunderstanding, but if we end up using those functions, shouldn't we just return Ids instead of objects?

File: backend/testUtils/testSession.ts (hunk 4 above)
Comment (mahadzaryab1) [reply backend/testUtils/testSession.ts]: oops I made a typo - fixed now!

File: backend/services/implementations/testSessionService.ts (hunk 5 above)
Comment (mahadzaryab1) [reply backend/services/implementations/testSessionService.ts:60]: think it should be fine - would probably become a problem if we were querying millions of documents

File: backend/services/interfaces/testSessionService.ts (hunk 6 above)
Comment (mahadzaryab1) [reply backend/services/interfaces/testSessionService.ts:53]: just to keep the request and response objects independent in case they need to diverge in the future

File: backend/testUtils/testSession.ts (hunk 3 above)
Comment (carissa-tang) [reply backend/testUtils/testSession.ts:31]: @joyce-shi within `createTestSession` we need to validate the 3 given ids and so we were thinking if we are already making the call, we can map the objects to the ids in the response. is there a downside you are considering with this approach?

File: backend/testUtils/testSession.ts (hunk 4 above)
Comment (carissa-tang) [reply backend/testUtils/testSession.ts]: i think the validation for this would be within `createTestResult`?

Comment (mahadzaryab1) [reply backend/testUtils/testSession.ts]: yep exactly @carissa-tang

File: backend/services/implementations/testSessionService.ts (hunk 5 above)
Comment (joyce-shi) [reply backend/services/implementations/testSessionService.ts:60]: cool, sounds good

File: backend/services/interfaces/testSessionService.ts (hunk 6 above)
Comment (joyce-shi) [reply backend/services/interfaces/testSessionService.ts:53]: makes sense, good call

File: backend/testUtils/testSession.ts (hunk 3 above)
Comment (joyce-shi) [reply backend/testUtils/testSession.ts:31]: oh shoot, i forgot about this. makes sense! sounds good.

**Review by joyce-shi — APPROVED** [review]
LGTM 🛳
